fix/record-collection (#36002)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/basic-javascript/record-collection.english.md

```yml
tests:

>   - text: Do not modify the <code>collection</code> object

Reviewer: I am not sure what verbiage should be here, but when I read this, it makes me think I need to return a copy of collection vs being able to modify the actual object. Maybe something like:
```suggestion
  - text: You should not change the original <code>collection</code> object's initialization.
```

>     testString: assert(code.match(/var collection = {\s*2548[:] {\s*"album"[:] "Slippery When Wet",\s*"artist"[:] "Bon Jovi",\s*"tracks"[:] \[\s*"Let It Rock",\s*"You Give Love a Bad Name"\s*\]\s*},\s*2468[:] {\s*"album"[:] "1999",\s*"artist"[:] "Prince",\s*"tracks"[:] \[\s*"1999",\s*"Little Red Corvette"\s*\]\s*},\s*1245[:] {\s*"artist"[:] "Robert Palmer",\s*"tracks"[:] \[ \]\s*},\s*5439[:] {\s*"album"[:] "ABBA Gold"\s*}\s*};/g));

Reviewer: What you have here basically works and I like the direction you took.  However,  I think it might be more maintainable to just add the following to the After Test section:
```
var collectionCopy = {
  2548: {
    "album": "Slippery When Wet",
    "artist": "Bon Jovi",
    "tracks": [
      "Let It Rock",
      "You Give Love a Bad Name"
    ]
  },
  2468: {
    "album": "1999",
    "artist": "Prince",
    "tracks": [
      "1999",
      "Little Red Corvette"
    ]
  },
  1245: {
    "artist": "Robert Palmer",
    "tracks": [ ]
  },
  5439: {
    "album": "ABBA Gold"
  }
};
```

Then, the testString simply becomes:
```suggestion
    testString: assert(JSON.stringify(collectionCopy) === JSON.stringify(collection));
```

The above would avoids the complex regex.

Author: Thanks for the review @RandellDawson 

I like the method here - but I'm not sure how to make it work - the function in the seed code modifies the original object when the tests are ran: 
```
// Alter values below to test your code
updateRecords(5439, "artist", "ABBA");
```
So it adds that record to the original object - and the users are supposed to use it to test their code, so they can change it to anything. So I'm not sure how to check with this method without removing that... or something - any ideas?

Also, do you think I should remove the quotes around the property names?

Reviewer: @moT01 OK, after much testing, I concur that since the camper can add extra calls to test (for the camper's own purposes) and these would alter the final testing, for now it is best to go with a regex version like you first attempted.  

I was wondering if you would be OK with a simplified version of the `testString` like the following? It is possible now that you have removed the quotation marks from the property names:
```
    testString: assert(code.match(/var collection = {\s*2548: {\s*album: "Slippery When Wet",\s*artist: "Bon Jovi",\s*tracks: \[\s*"Let It Rock",\s*"You Give Love a Bad Name"\s*\]\s*},\s*2468: {\s*album: "1999",\s*artist: "Prince",\s*tracks: \[\s*"1999",\s*"Little Red Corvette"\s*\]\s*},\s*1245: {\s*artist: "Robert Palmer",\s*tracks: \[ \]\s*},\s*5439: {\s*album: "ABBA Gold"\s*}\s*};/g));
```

Reviewer: Would you like me to make a suggestion you can commit?

```yml
tests:
  - text: You should not change the <code>collection</code> object's initialization

>     testString: assert(code.match(/var collection = {\s*2548[:] {\s*album[:] "Slippery When Wet",\s*artist[:] "Bon Jovi",\s*tracks[:] \[\s*"Let It Rock",\s*"You Give Love a Bad Name"\s*\]\s*},\s*2468[:] {\s*album[:] "1999",\s*artist[:] "Prince",\s*tracks[:] \[\s*"1999",\s*"Little Red Corvette"\s*\]\s*},\s*1245[:] {\s*artist[:] "Robert Palmer",\s*tracks[:] \[ \]\s*},\s*5439[:] {\s*album[:] "ABBA Gold"\s*}\s*};/g));

Reviewer: ```suggestion
    testString: assert(code.match(/var collection = {\s*2548: {\s*album: "Slippery When Wet",\s*artist: "Bon Jovi",\s*tracks: \[\s*"Let It Rock",\s*"You Give Love a Bad Name"\s*\]\s*},\s*2468: {\s*album: "1999",\s*artist: "Prince",\s*tracks: \[\s*"1999",\s*"Little Red Corvette"\s*\]\s*},\s*1245: {\s*artist: "Robert Palmer",\s*tracks: \[ \]\s*},\s*5439: {\s*album: "ABBA Gold"\s*}\s*};/g));
```

Author: Are you sure? I added them in because it wasn't working for me without them.

Reviewer: I am fairly sure.  Colon are not special characters that I am aware of.  Copy/paste what I posted above and see if it works with your latest changes.

Author: definitely getting an error in two ways...
- tried to save the file with these changes - then seed = error
- save a working file - npm run develop - change file to this + save = error
might just be me - but definitely some errors

Reviewer: I know the issue.  You have to surround the testString with single quotes to escape the colon in the yml.  I have edited my suggestion above.

Author: Single quotes looks like it worked for me - added to the PR
